Feature/#15025691 resize pencil erase

rachidcalazans opened the pull request:
## Why is this change necessary?
Because now we need to add new size of pencilSizeView tools and eraserSizeView tools.

## How does it address the issue?
Added 6 sizes for pencil and erase tools views.

## What side effects does this change have?
No side effects

Commit:
feature/#15025691 Update BrushTools

- add new size for pencil tool

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- update first size of pencil sizes
- add view and separator for size pencil view

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolsView

- update sizes for size pencil group

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- update position and icons for pencil size group

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- Add new subview group

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolVieww

- update new group positions

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- update way to add Pencil Group View

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- Update way to build erase tools
- add all sizes as pencil
- update the click colors for erase

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Add UI tests for Example project

- add UI tests for eraseGroupSize and pencilGroupSize
- update buttons to have identifier names

Signed-off-by: Stant Macmini n04 <stant_04@Mac-mini-de-Ralph.local>

Commit:
feature/#15025691 Update example Podfile

- change to use locally path

Signed-off-by: Stant Macmini n04 <stant_04@Mac-mini-de-Ralph.local>

Commit:
feature/#15025691 Update example ViewController

- add missing functions from protocols

Signed-off-by: Stant Macmini n04 <stant_04@Mac-mini-de-Ralph.local>

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -66,15 +77,15 @@ public class BrushToolView: UIView {
             setBtnAsClicked(button: uiButton)
         }
         
+        let y = (heightBaseToCenter - groupPencilSizeView.frame.height) / 2
```
Try always to align equalties.

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -126,40 +137,63 @@ public class BrushToolView: UIView {
-        btnSize3.uiButton.frame.size = CGSize(width: 28, height: 28)
-        btnSize4.uiButton.frame.size = CGSize(width: 33, height: 33)
-        btnSize5.uiButton.frame.size = CGSize(width: 37, height: 37)
+        let btnSize1   = Button(imageName: "pencilSize1", imageColor: blackHexColor, selector: #selector(self.pencilSize))
```
I know that wasn't your task, but try to rename for button ;)

rachidcalazans replied:
Changed :)

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -126,40 +137,63 @@ public class BrushToolView: UIView {
-        
-        groupEraseSizeButtons.append(eraser)
-        groupEraseSizeButtons.append(eraserFull)
+        let btnSize1   = Button(imageName: "eraserSize1", imageColor: blackHexColor, selector: #selector(self.eraserSize))
```
I know that wasn't your task, but try to rename for button ;)

rachidcalazans replied:
Done

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        
+        let originIconImg = CGPoint(x: (groupViewWidth - 25) / 2, y: 18)
+        let sizeIconImg   = CGSize(width: 25, height: 25)
+        let frameIconImg  = CGRect.init(origin: originIconImg, size: sizeIconImg)
```
Try to rename for "somethingIconImage", img is so sad ... hehe

rachidcalazans replied:
Done

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        
+        let pencilIconImageView = UIImageView(image: UIImage(named: "pencil", in: bundle, compatibleWith: nil))
+        
+        pencilIconImageView.accessibilityIdentifier = "pencilSizeReferenceIdentifier"
```
Extract identifiers for a separated method.

rachidcalazans replied:
Done

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let sizeSeparator   = CGSize(width: groupViewWidth - 20, height: 1)
+        let frameSeparator  = CGRect.init(origin: originSeparator, size: sizeSeparator)
+
+        separatorView.frame = frameSeparator
```
Identation.

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        groupPencilSizeView.isHidden   = true
+        
+        groupPencilSizeView.addSubview(pencilIconView)
+        groupPencilSizeView.addSubview(separatorView)
```
I think this method is so big that need some comments. Try to extract every function to a specific method, like groupPencilSizeViewConfiguration() that groups from line 269 to 276.

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let frameSeparator  = CGRect.init(origin: originSeparator, size: sizeSeparator)
+        
+        separatorView.frame = frameSeparator
+        separatorView.backgroundColor = UIColor.black
```
Identation.

liuvasconcelos commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        groupEraseSizeView.isHidden   = true
+        
+        groupEraseSizeView.addSubview(eraseIconView)
+        groupEraseSizeView.addSubview(separatorView)
```
Same above comment about code organization.

liuvasconcelos commented on example/MyFirstDrawing/MyFirstDrawing/ViewController.swift:
```diff
@@ -32,6 +32,13 @@ class ViewController: UIViewController {
 extension ViewController: DrawingDelegate {
+    func save(drawingImage: UIImage, drawingColor: String) {
+        print("Saved with drawingColor")
+    }
```
Does those method to anything?

rachidcalazans replied:
The DrawingDelegate Protocol had these new methods. I just added them. Demo implementation needs to do nothing.

liuvasconcelos commented on example/MyFirstDrawing/MyFirstDrawingUITests/BrushToolView/GroupEraseSizeViewUITests.swift:
```diff
@@ -0,0 +1,42 @@
+        btnErase.tap()
+    }
+    
+    override func tearDown() {}
```
If you are not doing anything here, I think that you could delete it.

rachidcalazans replied:
Done

liuvasconcelos commented on example/MyFirstDrawing/MyFirstDrawingUITests/BrushToolView/GroupEraseSizeViewUITests.swift:
```diff
@@ -0,0 +1,42 @@
+        let eraseSizeReference = app.images.matching(identifier: "eraserfullReferenceIdentifier")
+        
+        let btnEraseSize1 = app.buttons["eraserSize1Identifier"]
+        let btnEraseSize2 = app.buttons["eraserSize2Identifier"]
```
I was taking a look on this buttons and I was wondering if it was better to separate them to a different class. 
I'm not sure. It is better if there is any possibility for reuse of it. What do you think? 

--> On bigger apps, we decided to separate by pages.

rachidcalazans replied:
For this version, I don't believe it is necessary. Is just a small test with your setups. If the tests get bigger and need to reuse those options, we can extract it.

liuvasconcelos commented on example/MyFirstDrawing/MyFirstDrawingUITests/BrushToolView/GroupEraseSizeViewUITests.swift:
```diff
@@ -0,0 +1,42 @@
+        let btnEraseSize1 = app.buttons["eraserSize1Identifier"]
+        let btnEraseSize2 = app.buttons["eraserSize2Identifier"]
+        let btnEraseSize3 = app.buttons["eraserSize3Identifier"]
+        let btnEraseSize4 = app.buttons["eraserSize4Identifier"]
```
Anyway, avoid btn uses instead of button.

rachidcalazans replied:
Done

liuvasconcelos commented on example/MyFirstDrawing/MyFirstDrawingUITests/BrushToolView/GroupPencilSizeViewUITests.swift:
```diff
@@ -0,0 +1,42 @@
+        
+        let btnPencilSize1 = app.buttons["pencilSize1Identifier"]
+        let btnPencilSize2 = app.buttons["pencilSize2Identifier"]
+        let btnPencilSize3 = app.buttons["pencilSize3Identifier"]
```
Same comment above. If we decided to change over there, change it here as well.

rachidcalazans replied:
Done

liuvasconcelos commented on example/MyFirstDrawing/Podfile:
```diff
@@ -4,9 +4,8 @@
-
-  # Pods for MyFirstDrawing
 
+  pod 'StantDrawing', path: '../../StantDrawing.podspec'
```
Very nice!

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -66,15 +77,15 @@ public class BrushToolView: UIView {
         }
         
+        let y = (heightBaseToCenter - groupPencilSizeView.frame.height) / 2
+        let point = CGPoint(x: x, y: y)
```
As you said before, this function (addAsSubViewOn) is also used in other context. Does this change have impacts elsewhere?

rachidcalazans replied:
Yes, but I refactored the code to remove those lines as we talked before.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let frameIconView  = CGRect.init(origin: originIconView, size: sizeIconView)
+        
+        pencilIconView.frame              = frameIconView
+        pencilIconView.backgroundColor    = UIColor(hex: "F0F0F0")
```
This color could be extracted to a extension, because if any changes in the future are needed, you would only change in the extension, not on where you are using it.

rachidcalazans replied:
Like we talked. I extracted it to a let for this View.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        
+        setAllGroupdPencilSizeButtonsAsNotClicked()
+        
+        setBtnSizeAsClicked(button: groupPencilSizeButtons[1].uiButton)
```
Why is the second button size being set as clicked?

rachidcalazans replied:
To follow the default before of Sizes. We added a new size 1 with the half of size of the size 2.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let frameSeparator  = CGRect.init(origin: originSeparator, size: sizeSeparator)
+
+        separatorView.frame = frameSeparator
+        separatorView.backgroundColor = UIColor.black
```
Indentation here

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let sizeIconImg   = CGSize(width: 25, height: 25)
+        let frameIconImg  = CGRect.init(origin: originIconImg, size: sizeIconImg)
+        
+        let pencilIconImageView = UIImageView(image: UIImage(named: "pencil", in: bundle, compatibleWith: nil))
```
This would follow the same idea that I stated before about the color extension (in this case, would be an image extension).

rachidcalazans replied:
I don't believe in this case is necessary. We have only 1 line using it.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        // Separator
+        let separatorView   = UIView()
+        let originSeparator = CGPoint(x: 10, y: pencilIconView.frame.origin.y + pencilIconView.frame.height + 10)
+        let sizeSeparator   = CGSize(width: groupViewWidth - 20, height: 1)
```
Why does this - 20 mean? Can we avoid it by specifying the reason?

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
-        }
+        // Separator
+        let separatorView   = UIView()
+        let originSeparator = CGPoint(x: 10, y: pencilIconView.frame.origin.y + pencilIconView.frame.height + 10)
```
Why does this + 10 mean? Can we avoid it by specifying the reason?

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+
+        // Group Pencil Buttons
+        let groupPencilBtns = UIView()
+        let origin          = CGPoint(x: 0, y: separatorView.frame.origin.y + separatorView.frame.height + 10)
```
Why does this + 10 mean? Can we avoid it by specifying the reason?

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let frameIconView  = CGRect.init(origin: originIconView, size: sizeIconView)
+        
+        eraseIconView.frame              = frameIconView
+        eraseIconView.backgroundColor    = UIColor(hex: "F0F0F0")
```
Same comment as stated before about colors extensions.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        let sizeIconImg   = CGSize(width: 25, height: 25)
+        let frameIconImg  = CGRect.init(origin: originIconImg, size: sizeIconImg)
+        
+        let eraseIconImageView = UIImageView(image: UIImage(named: "eraserfull", in: bundle, compatibleWith: nil))
```
Same comment as stated before about images extensions.

rachidcalazans replied:
I don't believe in this case is necessary. We have only 1 line using it.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        
+        // Separator
+        let separatorView   = UIView()
+        let originSeparator = CGPoint(x: 10, y: eraseIconView.frame.origin.y + eraseIconView.frame.height + 10)
```
Same comments as stated before about magic numbers.

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        separatorView.alpha = 0.2
+        
+        // Group Erase Buttons
+        let groupEraseBtns = UIView()
```
Alignment

alexandrehcdc commented on StantDrawing/BrushTool/BrushToolView.swift:
```diff
@@ -187,27 +221,115 @@ public class BrushToolView: UIView {
+        setBtnSizeAsClicked(button: groupEraseSizeButtons[1].uiButton)
+        
+        // GroupPencilSizeView Configs
+        let height    = eraseIconView.frame.size.height + separatorView.frame.size.height + groupEraseBtns.frame.size.height
```
Block alignment

alexandrehcdc commented on example/MyFirstDrawing/MyFirstDrawingUITests/BrushToolView/GroupEraseSizeViewUITests.swift:
```diff
@@ -0,0 +1,42 @@
+        btnErase.tap()
+    }
+    
+    override func tearDown() {}
```
Is it ok if you don't have any code here on tearDown? Specially for actions like app.launch(), does it have a impact on tests or it is ok to leave it like that?

rachidcalazans replied:
It's not necessary. I've removed

Commit:
feature/#15025691 Update BrushToolView

- refactor way to put the GroupSize and coolors on the view
- remove unused funcs on tests

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- extract a UIColor default for backgroundGroup

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- update the aligments

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- update bad names
- refactor way to config groupSizeViews

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update BrushToolView

- replace old funcs to the new to setAsNotClicked buttons

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update DrawingView

- update default size for erase

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>

Commit:
feature/#15025691 Update UITests names

- update bad names

Signed-off-by: Stant Macmini n04 <stant_04@Mini-de-Ralph.stant.local>
